lib/Doctrine/ORM/Tools/Pagination/RemoveUselessLeftJoinsWalker.php

```diff
@@ -0,0 +1,150 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Doctrine\ORM\Tools\Pagination;
+
+use Doctrine\ORM\Query;
+use Doctrine\ORM\Query\AST\SelectStatement;
+use Doctrine\ORM\Query\TreeWalkerAdapter;
+use function array_filter;
+use function array_key_exists;
+use function array_merge;
+use function array_unique;
+use function in_array;
+use function reset;
+
+class RemoveUselessLeftJoinsWalker extends TreeWalkerAdapter
+{
+    public function walkSelectStatement(SelectStatement $AST)
+    {
+        $this->removeUnusedJoins($AST);
+    }
+
+    private function removeUnusedJoins(SelectStatement $AST)
```

Comment:
This feature takes care only about joins in subselect? Method should be renamed then, probably also whole class.